### ISSUE TITLE
add function to return last date of week

### ISSUE THREAD

**Issue by mbtolou** [ClickHouse/ClickHouse#43908]
Hi.

Is there a way to get the closest Friday or Saturday to the received date?

function like `toFriday` or 'toSaturday' or `toNearestDayOfWeek` .

In many Asian countries, Friday is the last day of the week and Saturday is the first day of the week.

**Comment by den-crane** [comment]
WA:

```sql
create or replace function toFriday as (day) -> 
  toDayOfWeek(day) >= 5 ? addDays(toMonday(day),4) : subtractDays(addDays(toMonday(day),4), 7);


SELECT
    toDate('2022-12-01') + number AS d,
    toDayOfWeek(d) = 5 AS isFriday,
    toFriday(d) AS toFriday
FROM numbers(20)


┌──────────d─┬─isFriday─┬───toFriday─┐
│ 2022-12-01 │        0 │ 2022-11-25 │
│ 2022-12-02 │        1 │ 2022-12-02 │
│ 2022-12-03 │        0 │ 2022-12-02 │
│ 2022-12-04 │        0 │ 2022-12-02 │
│ 2022-12-05 │        0 │ 2022-12-02 │
│ 2022-12-06 │        0 │ 2022-12-02 │
│ 2022-12-07 │        0 │ 2022-12-02 │
│ 2022-12-08 │        0 │ 2022-12-02 │
│ 2022-12-09 │        1 │ 2022-12-09 │
│ 2022-12-10 │        0 │ 2022-12-09 │
│ 2022-12-11 │        0 │ 2022-12-09 │
│ 2022-12-12 │        0 │ 2022-12-09 │
│ 2022-12-13 │        0 │ 2022-12-09 │
│ 2022-12-14 │        0 │ 2022-12-09 │
│ 2022-12-15 │        0 │ 2022-12-09 │
│ 2022-12-16 │        1 │ 2022-12-16 │
│ 2022-12-17 │        0 │ 2022-12-16 │
│ 2022-12-18 │        0 │ 2022-12-16 │
│ 2022-12-19 │        0 │ 2022-12-16 │
│ 2022-12-20 │        0 │ 2022-12-16 │
└────────────┴──────────┴────────────┘
```

**Comment by mbtolou** [comment]
Thanks.
Can I use this function for partitioning table ?


```sql
CREATE TABLE
	test.signal_logs (
		`clock` DateTime ('Asia/Tehran'),
		`device_id` LowCardinality (String),
		`signal` UInt8,
		`r_time` Nullable (DateTime ('Asia/Tehran')) DEFAULT now ('Asia/Tehran') COMMENT 'receive time'
	) ENGINE = MergeTree
PARTITION BY
	toFriday (clock)
ORDER BY
	(device_id, clock) SETTINGS index_granularity = 8192;
```

**Comment by rvasin** [comment]
This function is implemented in #50315. But the issue should be open (it's correct). Because  toLastDayOfWeek()  should have an additional argument to specify which day of week is start of week.
